### PR TITLE
llext: Add RISC-V arch-specific relocations

### PR DESCRIPTION
**Pull request by WorldofJARcraft**
This commit introduces architecture-specific ELF relocations for RISC-V, in accordance with the RISC-V PSABI specification: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc.
Also, the necessary compiler configurations for compiling LLEXT extensions on RISC-V are added, and the llext tests are executed on RISC-V targets.
Calling llext extensions from user threads in RISC-V is still unsupported as of this commit.

### CODE REVIEW

File: subsys/llext/CMakeLists.txt [hunk 1]
```diff
@@ -17,4 +17,8 @@ if(CONFIG_LLEXT)
     fs_loader.c
 	)
   zephyr_library_sources_ifdef(CONFIG_LLEXT_SHELL shell.c)
+
+  if(CONFIG_RISCV AND CONFIG_USERSPACE)
+	message(WARNING "Running LLEXT extensions from user-space threads on RISC-V is not supported!")
```
Comment (pillo79) [inline subsys/llext/CMakeLists.txt:22]: Should this be a FATAL_ERROR, or can you still run llexts in supervisor mode?
Please add a RISCV section in `doc/services/llext` with these details / caveats.

File: subsys/llext/llext_load.c
Comment (pillo79) [inline subsys/llext/llext_load.c:1]: This file looks automatically formatted - most of the changed lines were indented for readability though.
Can you avoid reformatting the whole file? At the very least, please make a cleanup commit that applies the new style but does no functional changes, so it is easier to spot the actual PR code.

File: subsys/llext/llext_load.c [hunk 2]
```diff
@@ -390,14 +383,37 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext)
...
+			 */
+			if (IS_ENABLED(CONFIG_RISCV) && i == LLEXT_MEM_RODATA &&
+			    j == LLEXT_MEM_DATA) {
+				continue;
+			}
+
+			if (IS_ENABLED(CONFIG_RISCV) && i == LLEXT_MEM_DATA &&
+			    j == LLEXT_MEM_RODATA) {
+				continue;
+			}
```
Comment (pillo79) [inline subsys/llext/llext_load.c:407]: Not sure it's OK to ignore this overlap - depending on how RISCV manages the memory partitions, the end result might be read-only access to `.sdata`. 
Can the ordering be fixed with a linker script?

File: tests/subsys/llext/simple/src/test_llext_simple.c [hunk 3]
```diff
@@ -417,9 +417,12 @@ ZTEST(llext, test_find_section)
 	uintptr_t symbol_ptr = (uintptr_t)llext_find_sym(&ext->exp_tab, "number");
 	uintptr_t section_ptr = (uintptr_t)find_section_ext + section_ofs;
 
-	zassert_equal(symbol_ptr, section_ptr,
+	/* on RISC-V, at least for GCC, the symbols aren't always at the beginning of the section */
+	if (IS_ENABLED(CONFIG_RISCV)) {
+		zassert_equal(symbol_ptr, section_ptr + section_padding,
```
Comment (pillo79) [inline tests/subsys/llext/simple/src/test_llext_simple.c:422]: This now checks RISCV  _only_... :slightly_smiling_face: 
Also, `section_padding` looks undefined. Maybe have it default to 0 for non-RISCV targets?

**Review by pillo79 — COMMENTED** [review]
This is great stuff, thanks! Here's some initial feedback before an in-depth review.

File: subsys/llext/CMakeLists.txt (hunk 1 above)
Comment (WorldofJARcraft) [reply subsys/llext/CMakeLists.txt:22]: So far, I have not encountered any issues running llexts in supervisor mode.
I added this warning because I encountered an instruction access fault in the `llext.simple.readonly_mpu` scenario in the llext test suite, specifically when attempting to run the llext in a user thread. This appears to be caused by incorrect configuration of the PMP, which is probably unrelated to the relocations.

File: subsys/llext/llext_load.c
Comment (WorldofJARcraft) [reply subsys/llext/llext_load.c:1]: I will undo the formatting changes.

File: tests/subsys/llext/simple/src/test_llext_simple.c (hunk 3 above)
Comment (WorldofJARcraft) [reply tests/subsys/llext/simple/src/test_llext_simple.c:422]: You are right, I wanted to test this on everything BUT RISC-V. Also, I will remove the padding.

File: subsys/llext/llext_load.c (hunk 2 above)
Comment (WorldofJARcraft) [reply subsys/llext/llext_load.c:407]: It should be possible to re-order the sections in a linker script. 
I am not sure why this scenario only occurs on RISC-V, to be honest.

File: subsys/llext/llext_load.c
Comment (WorldofJARcraft) [reply subsys/llext/llext_load.c:1]: I have reverted the changed formatting, leaving only the actual code changes.

File: subsys/llext/llext_load.c [hunk 4]
```diff
@@ -365,6 +365,29 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext)
 				continue;
 			}
 
+			/*
+			 * Exported symbols region can also overlap
+			 * with rodata.
+			 */
+			if (i == LLEXT_MEM_EXPORT || j == LLEXT_MEM_EXPORT) {
+				continue;
```
Comment (WorldofJARcraft) [inline subsys/llext/llext_load.c:373]: I found this using a (fairly complex) example of my own, that is currently out-of-tree; it might not be necessary to include this to pass the current unit tests.

File: subsys/llext/llext_load.c (hunk 2 above)
Comment (WorldofJARcraft) [reply subsys/llext/llext_load.c:389]: Maybe this an idiosyncrasy of my specific compiler? I am using the pre-compiled zephyr sdk (zephyr-sdk-0.16.5-1).

File: subsys/llext/llext_load.c (hunk 4 above)
Comment (pillo79) [reply subsys/llext/llext_load.c:373]: LGTM, makes sense to keep it. I have a similar exception in one of my branches as well :slightly_smiling_face:

**Review by pillo79 — COMMENTED** [review]
LGTM; only thing I would rather improve is the `DATA`/`RODATA` overlap silencing.

**Comment by WorldofJARcraft** [comment]
> LGTM; only thing I would rather improve is the `DATA`/`RODATA` overlap silencing.

I looked into the cause of this issue a bit more. The issue only manifests itself with the `LLEXT_TYPE_ELF_OBJECT` binary type, i.e., a binary created using gcc -c. I have used an overwrite in the test case to force `LLEXT_TYPE_ELF_RELOCATABLE` instead, and I did not encounter any issues.
Here are three ideas for fixes that would allow me to remove the overlap silencing, do you have a preference?
1. Prohibit the `LLEXT_TYPE_ELF_OBJECT` on RISC-V in Kconfig.
2. (only on RISC-V) run the `LLEXT_TYPE_ELF_OBJECT` through a basic linker pass.
3. (only on RISC-V) disable use of the sections that cause the perceived conflict when `LLEXT_TYPE_ELF_OBJECT` is enabled using gcc options.

File: tests/subsys/llext/simple/src/test_llext_simple.c (hunk 3 above)
Comment (WorldofJARcraft) [reply tests/subsys/llext/simple/src/test_llext_simple.c]: This problem also only occured with `LLEXT_TYPE_ELF_OBJECT`; I have removed the condition and added a comment stating that this will break if support for `LLEXT_TYPE_ELF_OBJECT` is ever added again.

**Comment by teburd** [comment]
> > Prohibit the `LLEXT_TYPE_ELF_OBJECT` on RISC-V in Kconfig.
> 
> IMHO this (along with making `RELOCATABLE` the default for RISC-V), is the easiest of the three and has minimal impact. `OBJECT` is basically a legacy choice nowadays; `RELOCATABLE` is the same as `OBJECT` + an extra linker pass. If you have only one file, they are very similar. Unless your toolchain and/or target has issues with `-r`... 😅

If you are looking to use the relocatable type (partially linked) then I agree with this.

Does RISC-V have program counter relative jumps? If so, and the jump is too far away this should currently be an error. I didn't look too closely at the code yet.

**Comment by WorldofJARcraft** [comment]
TL;DR: You are right, I will add checks accordingly and get back to you.

For 32-bit RISC-V, compilers commonly generate pairs of `auipc` or `lui` (U-type, 20-bit immediate) and an I-type instruction like `jr` (I-type, 12-bit immediate). Thereby, any absolute or PC-relative address in the 32-bit address space can be reached. Which instructions are generated depends on the code model (`auipc` for medany, `lui` for medlow, if I remember correctly - see the specification I commented in the header file).
For 64-bit RISC-V, the same code models can be used to cover a 32-bit range around the PC or the lowest 2^32 bytes, respectively. There is also the large code model, where the compiler generates a table before the beginning of each function, where the linker is expected to place 64-bit addresses for remote symbols. Thereby, any 64-bit address is reachable.

**Comment by WorldofJARcraft** [comment]
> TL;DR: You are right, I will add checks accordingly and get back to you.
> 
> For 32-bit RISC-V, compilers commonly generate pairs of `auipc` or `lui` (U-type, 20-bit immediate) and an I-type instruction like `jr` (I-type, 12-bit immediate). Thereby, any absolute or PC-relative address in the 32-bit address space can be reached. Which instructions are generated depends on the code model (`auipc` for medany, `lui` for medlow, if I remember correctly - see the specification I commented in the header file). For 64-bit RISC-V, the same code models can be used to cover a 32-bit range around the PC or the lowest 2^32 bytes, respectively. There is also the large code model, where the compiler generates a table before the beginning of each function, where the linker is expected to place 64-bit addresses for remote symbols. Thereby, any 64-bit address is reachable.

I have added checks whether the relocations fit.

**Review by teburd — DISMISSED** [review]
This looks great, I doubt we've covered all the relocations in our tests and it would be great to add some riscv specific testing to ensure we cover all the relocations supported in riscv's elf.c

**Comment by WorldofJARcraft** [comment]
Some RISC-V specific edge cases require a very new compiler; e.g., the large code model and the related relocations are only supported in GCC 14.0+. This might cause false positives in the CI.

File: tests/subsys/llext/simple/testcase.yaml
Comment (pillo79) [inline tests/subsys/llext/simple/testcase.yaml:1]: Minor nit/question: are all the new `RELOCATABLE` lines here required? Unless the test requires the file type to be forced, it should just pick the arch default, right?

Comment (teburd) [reply tests/subsys/llext/simple/testcase.yaml:1]: yeah great point

Comment (WorldofJARcraft) [reply tests/subsys/llext/simple/testcase.yaml:1]: Changed accordingly.

**Comment by pillo79** [comment]
@WorldofJARcraft any updates on the CI issues? I'd really love to see this in!

**Comment by WorldofJARcraft** [comment]
The failing tests seem to be unrelated to the changes. I will re-base and try again, maybe this will fix it.

**Comment by WorldofJARcraft** [comment]
I have re-based my PR to main.
All llext tests pass locally when I run 
```bash
./scripts/twister -v -T tests/subsys/llext/simple/
```
This includes the tests that failed in the last CI run.
Looking through the logs from the last failing CI run, I noticed differences in the llext ELF files between the runs in the CI and my local runs (different section sizes etc.), so this might be a toolchain issue. I am using the latest zephyr toolchain (zephyr-sdk-0.16.8). 
If twister fails again in the CI, I will try again to reproduce the issue locally. However, I am not sure how to approach this.

**Review by teburd — APPROVED** [review]

**Review by pillo79 — APPROVED** [review]
